[Enhancement] - Incident mirroring and Incident Structure update for Microsoft 365 Defender (#37480)
demisto/content: Packs/Microsoft365Defender/Integrations/Microsoft365Defender/README.md

   3. Under **Incident type**, select Microsoft 365 Defender Incident. 
   4. Under **Mapper (incoming)**, select Microsoft 365 Defender - Incoming Mapper.
   5. Under **Mapper (outgoing)**, select Microsoft 365 Defender - Outgoing Mapper.
   6. To enable mirroring to close a ticket in Cortex XSOAR, check the **Close Mirrored Cortex XSOAR Incidents** checkbox.
   7. To enable mirroring to close an incident in Microsoft 365 Defender, check the **Close Mirrored Microsoft 365 Defender Incidents** checkbox.

| **Parameter** | **Description** | **Required** |
| --- | --- | --- |
| Endpoint URI | The United States: api-us.security.microsoft.com<br/>Europe: api-eu.security.microsoft.com<br/>The United Kingdom: api-uk.security.microsoft.co | True |
| ID or Client ID |  | False |
| Token or Tenant ID |  | False |
| Application ID | The API key to use to connect. | False |
| Use Client Credentials Authorization Flow | Use a self-deployed Azure application and authenticate using the Client Credentials flow. | False |
| Tenant ID (for Client Credentials mode) |  | False |
| Client Secret (for Client Credentials mode) |  | False |
| Client Secret |  | False |

> | Certificate Thumbprint | Used for certificate authentication. As appears in the "Certificates &amp;amp; secrets" page of the app. | False |

Reviewer: ```suggestion
| Certificate Thumbprint | Used for certificate authentication. As appears in the "Certificates &amp; secrets" page of the app. | False |
```

| First fetch timestamp (&lt;number&gt; &lt;time unit&gt;, e.g., 12 hours, 7 days) |  | False |
| Fetch incidents timeout | The time limit in seconds for fetch incidents to run. Leave this empty to cancel the timeout limit. | False |
| Number of incidents for each fetch. | Due to API limitations, the maximum is 100. | False |
| Incident type |  | False |
| Fetch incidents |  | False |
| Trust any certificate (not secure) |  | False |
| Use system proxy settings |  | False |
| Application ID (Deprecated) |  | False |
| Tenant ID (for Client Credentials mode) (Deprecated) |  | False |
| Client Secret (for Client Credentials mode) (Deprecated) |  | False |
| Incidents Fetch Interval |  | False |
| Incident Mirroring Direction | Choose the direction to mirror the incident: Incoming \(from Microsoft 365 Defender to Cortex XSOAR\), Outgoing \(from Cortex XSOAR to  Microsoft 365 Defender\), or Incoming and Outgoing \(from/to Cortex XSOAR and  Microsoft 365 Defender\). | False |
| Close Mirrored Cortex XSOAR Incidents | Incoming Mirroring - when selected, closing the Microsoft 365 Defender incident is mirrored in Cortex XSOAR. | False |

> | Close Mirrored Microsoft 365 Defender Incidents | Outgoing Mirroring - when selected, closing the XSOAR incident is mirrored in Microsoft 365 Defender. | False |

Reviewer: ```suggestion
| Close Mirrored Microsoft 365 Defender Incidents | Outgoing Mirroring - when selected, closing the Cortex XSOAR incident is mirrored in Microsoft 365 Defender. | False |
```

| Comment Entry Tag To Microsoft 365 Defender | Choose a tag to add to an entry to mirror it as a comment into Microsoft 365 Defender. | False |
| Comment Entry Tag From Microsoft 365 Defender | Choose a tag to add to an entry to mirror it as a comment from Microsoft 365 Defender. | False |

5. Run the !microsoft-365-defender-auth-test command to validate the authentication process.

## Incident Mirroring

> ### Mirroring In (Microsoft 365 Defender → XSOAR)

Reviewer: ```suggestion
### Mirroring In (Microsoft 365 Defender → Cortex XSOAR)
```

> When incidents are mirrored into XSOAR from Microsoft 365 Defender:

Reviewer: ```suggestion
When incidents are mirrored into Cortex XSOAR from Microsoft 365 Defender:
```

> 1. **Comments** mirrored from Microsoft 365 Defender will be added to the incident as entries in XSOAR and tagged with the **Comment Entry Tag From Microsoft 365 Defender**.  

Reviewer: ```suggestion
1. **Comments** mirrored from Microsoft 365 Defender will be added to the incident as entries in Cortex XSOAR and tagged with **Comment Entry Tag From Microsoft 365 Defender**.  
```

   * By default, the tag is set to: `CommentFromMicrosoft365Defender`.  

2. If an incident is closed in Microsoft 365 Defender (`status` = **Resolved**) and Close Mirrored Cortex XSOAR Incidents is enabled:  

>    * The **"classification"** field in Microsoft 365 Defender will be mapped to the **Close Reason** field in XSOAR.  

Reviewer: ```suggestion
   * The **"classification"** field in Microsoft 365 Defender will be mapped to the **Close Reason** field in Cortex XSOAR.  
```

> ### Mirroring Out (XSOAR → Microsoft 365 Defender)

Reviewer: ```suggestion
### Mirroring Out (Cortex XSOAR → Microsoft 365 Defender)
```

> When incidents are mirrored out from XSOAR to Microsoft 365 Defender:

Reviewer: ```suggestion
When incidents are mirrored out from Cortex XSOAR to Microsoft 365 Defender:
```

1. **Supported Fields**:  
   The following fields are mirrored:  
   * **Microsoft 365 Defender Status**  
   * **Assigned User**  
   * **Microsoft 365 Defender Classification**  
   * **Microsoft 365 Defender Tags**  

2. **Comments**:  

>    * XSOAR entries with the **Comment Entry Tag To Microsoft 365 Defender** tag are mirrored as comments in Microsoft 365 Defender.  

Reviewer: ```suggestion
   * Cortex XSOAR entries with the **Comment Entry Tag To Microsoft 365 Defender** tag are mirrored as comments in Microsoft 365 Defender.  
```

   * **Note**: Comments cannot be edited or deleted in Microsoft 365 Defender, as this functionality is not supported.

> 3. If an incident is closed in XSOAR and Close Mirrored Microsoft 365 Defender Incidents is enabled:  

Reviewer: ```suggestion
3. If an incident is closed in Cortex XSOAR and Close Mirrored Microsoft 365 Defender Incidents is enabled:  
```

   * The **Close Reason** field is mirrored to the `status` and `classification` fields in Microsoft 365 Defender.  

### Closing Logic

> #### Incoming Closing Logic (Microsoft 365 Defender → XSOAR)

Reviewer: ```suggestion
#### Incoming Closing Logic (Microsoft 365 Defender → Cortex XSOAR)
```

When an incident is resolved in Microsoft 365 Defender:

1. If the `status` = **Resolved**:  

>    * The **"classification"** field will be mapped to the **Close Reason** field in XSOAR using the **Close Reason Mapping Table**.  
>    * The **"classification"** and  **"determination"** fields will be mirrored into the **Microsoft 365 Defender Classification** field in XSOAR.

Reviewer: ```suggestion
   * The **"classification"** field will be mapped to the **Close Reason** field in Cortex XSOAR using the **Close Reason Mapping Table**.  
   * The **"classification"** and  **"determination"** fields will be mirrored into the **Microsoft 365 Defender Classification** field in Cortex XSOAR.
```

**Example**:  
* An incident closed in Microsoft 365 Defender with:  
  * `status` = **Resolved**
  * `Classification` = **True Positive**  
  * `Determination` = **Phishing**  

>   Results in XSOAR:  

Reviewer: ```suggestion
  Results in Cortex XSOAR:  
```

  * `Close Reason` = **Resolved**  
  * `Microsoft 365 Defender Classification` = **True Positive - Phishing**

**Close Reason Mapping Table (Incoming)**:

> | Classification in Microsoft 365 Defender | Close Reason in XSOAR      |

Reviewer: ```suggestion
| Classification in Microsoft 365 Defender | Close Reason in Cortex XSOAR      |
```

|-----------------------------------------|----------------------------|
| Not set                                 | Other                      |
| True Positive                           | Resolved                   |
| False Positive                          | False Positive             |
| Informational / Expected Activity       | Resolved                   |

> #### Outgoing Closing Logic (XSOAR → Microsoft 365 Defender)
> 
> When closing incidents from XSOAR to Microsoft 365 Defender, the following logic is applied to map the **Close Reason** and **Microsoft 365 Defender Classification** fields:
> 
> 1. **General Status Update**:  
>    When an incident is marked as closed in XSOAR, the `status` field in Microsoft 365 Defender will always be set to **Resolved**.  
> 
> 2. **Close Reason: Resolved**  
>   If the **Close Reason** in XSOAR is **Resolved**, the **Microsoft 365 Defender Classification** field is mirrored to classification and determination fields **as-is** into Microsoft 365 Defender without any changes.  
> 
> 3. **Close Reason: False Positive**  
>    If the **Close Reason** in XSOAR is **FalsePositive**, but the `classification` field does not already equal `FalsePositive`, the following updates occur:  
>    * **classification** → `FalsePositive`  
>    * **determination** → `Other`  
> 
> 3. **Close Reason: Other or Duplicate**  
>    If the **Close Reason** in XSOAR is either **Other** or **Duplicate**, the following updates occur:  
>    * **classification** → `Unknown`  
>    * **determination** → `NotAvailable`  
> 
> 
> **Close Reason Mapping Table (Outgoing)**:
> 
> | Close Reason in XSOAR | Classification in Microsoft 365 Defender | Determination in Microsoft 365 Defender | Status in Defender |

Reviewer: ```suggestion
#### Outgoing Closing Logic (Cortex XSOAR → Microsoft 365 Defender)

When closing incidents from Cortex XSOAR to Microsoft 365 Defender, the following logic is applied to map the **Close Reason** and **Microsoft 365 Defender Classification** fields:

1. **General Status Update**:  
   When an incident is marked as closed in Cortex XSOAR, the `status` field in Microsoft 365 Defender will always be set to **Resolved**.  

2. **Close Reason: Resolved**  
  If the **Close Reason** in Cortex XSOAR is **Resolved**, the **Microsoft 365 Defender Classification** field is mirrored to classification and determination fields **as-is** into Microsoft 365 Defender without any changes.  

3. **Close Reason: False Positive**  
   If the **Close Reason** in Cortex XSOAR is **FalsePositive**, but the `classification` field does not already equal `FalsePositive`, the following updates occur:  
   * **classification** → `FalsePositive`  
   * **determination** → `Other`  

3. **Close Reason: Other or Duplicate**  
   If the **Close Reason** in Cortex XSOAR is either **Other** or **Duplicate**, the following updates occur:  
   * **classification** → `Unknown`  
   * **determination** → `NotAvailable`  

**Close Reason Mapping Table (Outgoing)**:

| Close Reason in Cortex XSOAR | Classification in Microsoft 365 Defender | Determination in Microsoft 365 Defender | Status in Defender |
```

|-----------------------|------------------------------------------|-----------------------------------------|--------------------|
| Resolved              | Mirrored as-is                           | Mirrored as-is                          | Resolved           |
| FalsePositive         | FalsePositive                            | Mirrored as-is \ Other                  | Resolved           |
| Other                 | Unknown                                  | NotAvailable                            | Resolved           |
| Duplicate             | Unknown                                  | NotAvailable                            | Resolved           |

#### Notes

1. **Configuration Requirements for Closing Incidents**:  

>    To close incidents in both directions, ensure the following settings in XSOAR:  

Reviewer: ```suggestion
   To close incidents in both directions, ensure the following settings in Cortex XSOAR:  
```

   * **Close Mirrored XSOAR Incident** checkbox = **True**  
   * **Close Mirrored Microsoft 365 Defender Incident** checkbox = **True**  

   If these checkboxes are **False**, only the `Microsoft 365 Defender Status` and `Microsoft 365 Defender Classification` fields will be updated, and incidents will not be closed.

### Configure Incident Mirroring

> **This feature is compliant with XSOAR version 6.0 and above.**  

Reviewer: ```suggestion
**This feature is compliant with Cortex XSOAR version 6.0 and above.**  
```

When mirroring incidents, you can make changes in Microsoft 365 Defender that will be reflected in Cortex XSOAR, or vice versa. 

> Mirroring out of the box includes the following fields:

Reviewer: What fields?

The following instructions include steps for configuring the integration and incoming and outgoing mappers. However, they do not cover every option available in the integration nor classification and mapping features. 

> For information about classification and mapping see [Classification and Mapping](https://docs-cortex.paloaltonetworks.com/r/Cortex-XSOAR/6.10/Cortex-XSOAR-Administrator-Guide/Classification-and-Mapping).  

Reviewer: ```suggestion
For information about classification and mapping see:
- For Cortex XSOAR 6.13, see [Classification and Mapping](https://docs-cortex.paloaltonetworks.com/r/Cortex-XSOAR/6.13/Cortex-XSOAR-Administrator-Guide/Classification-and-Mapping).  
- For Cortex XSOAR 8 Cloud, see [Classification and Mapping](https://docs-cortex.paloaltonetworks.com/r/Cortex-XSOAR/8/Cortex-XSOAR-Cloud-Documentation/Classification-and-mapping). 
- For Cortex XSOAR 8.7 on-prem, see [Classification and Mapping](https://docs-cortex.paloaltonetworks.com/r/Cortex-XSOAR/8.7/Cortex-XSOAR-On-prem-Documentation/Classification-and-mapping). 
```
Is this applicable to Cortex XSOAR 8.7 on-prem?

**Note:**  
* For Cortex XSOAR version 6.1 only, the final source of truth for an incident are the values in Cortex XSOAR.  For example, if you change the severity in Cortex XSOAR and then change it back in Microsoft 365 Defender, the final value that will be presented is the one in Cortex XSOAR. For versions 6.2 and later, if mirroring is in both directions then the latest update is the source of truth. 
* The mirroring settings apply only for incidents that are fetched after applying the settings. 
* To use a custom mapper, you must first duplicate the mapper and edit the field in the copy of the mapper. If you detach the out of the box mapper and make changes to it, the pack does not automatically get updates.

#### STEP 1 - Configure the Microsoft 365 Defender Integration Instance for Mirroring.

1. Navigate to **Integrations** and search for ** Microsoft 365 Defender**.
2. Click **Add instance**.
3. Select **Fetches incidents**.

> 4. Select the **Incident Mirroring Direction**:  

Reviewer: ```suggestion
4. Select **Incident Mirroring Direction**:  
```

7. Under **Mapper (outgoing)**, for default mapping select Microsoft 365 Defender - Outgoing Mapper. For custom mapping, follow the instructions in STEP 3 and then select the custom mapper name.
8. Enter the relevant **Comment Entry Tag To Microsoft 365 Defender** and **Comment Entry Tag From Microsoft 365 Defender** values.  
These values are mapped to the **dbotMirrorTags** incident field in Cortex XSOAR, which defines how Cortex XSOAR handles comments when you tag them in the War Room. 
**Note:**  
These tags work only for mirroring comments from Cortex XSOAR to Microsoft 365 Defender.
9. To enable mirroring when closing an incident in Cortex XSOAR and Microsoft 365 Defender, select the **Close Mirrored Cortex XSOAR Incidents** and **Close Mirrored Microsoft 365 Defender Incidents** checkboxes respectively.
10. Click **Save & Exit**.

#### STEP 2 (Optional) Configure the Incoming Mapper by Incident Type

**Note:**
Any modifications require that the mappers be cloned before any changes can be applied.

> 1. Navigate to **Classification and Mapping** and for **Incidents** search for the **Microsoft 365 Defender - Incoming Mapper**.

Reviewer: ```suggestion
1. Navigate to **Classification and Mapping** and for **Incidents** search for **Microsoft 365 Defender - Incoming Mapper**.
```

> 2. Select it and click **Duplicate**.

Reviewer: ```suggestion
2. Select **Microsoft 365 Defender - Incoming Mapper** and click **Duplicate**.
```

> 3. Under the **Incident Type** dropdown, select Microsoft 365 Defender Incident.

Reviewer: ```suggestion
3. Under **Incident Type**, select **Microsoft 365 Defender Incident**.
```

4. Verify the mapper has these fields mapped. They will pull the values configured on the integration instance settings at the time of ingestion.
    * **dbotMirrorId** - dbotMirrorId - the field used by the third-party integration to identify the incident. This should be the incidentId of the Microsoft 365 Defender Incident.
    * **dbotMirrorDirection** - determines whether mirroring is incoming, outgoing, or both. Default is Both. This should match the instance configuration.        
    * **dbotMirrorInstance** - determines the Microsoft 365 Defender instance with which to mirror. This should match the instance configuration.
    * **dbotMirrorLastSync** - determines the field by which to indicate the last time that the systems synchronized.
    * **dbotMirrorTags** - determines the tags that you need to add in Cortex XSOAR for entries to be pushed to Microsoft 365 Defender. They should be copied from the tags in the instance configuration. These are also the tags that must be put on the War Room record in order for it to sync.

#### STEP 3 - Modify the Outgoing Mapper  

**Note:**  
Any modifications require that the mappers be cloned before any changes can be applied.

> 1. Navigate to **Classification and Mapping**, and for **Incidents** search for the **Microsoft 365 Defender - Outgoing Mapper.**
> 2. Select it and click **Duplicate**.  

Reviewer: ```suggestion
1. Navigate to **Classification and Mapping**, and for **Incidents** search for **Microsoft 365 Defender - Outgoing Mapper**.
2. Select **Microsoft 365 Defender - Outgoing Mapper** and click **Duplicate**.  
```

  The left side of the screen shows the Microsoft 365 Defender fields to which to map and the right side of the
screen shows the Cortex XSOAR fields by which you are mapping.

> 3. Under the **Incident Type** dropdown, select the relevant incident type (for example **Microsoft 365 Defender Incident**).

Reviewer: ```suggestion
3. Under **Incident Type** , select the relevant incident type (for example **Microsoft 365 Defender Incident**).
```

4. Under **Schema Type**, select **incident**. The Schema Type represents the Microsoft 365 Defender entity that
you are mapping to.
5. On the right side of the screen, under **Incident**, select the incident based on which you want to
match.
6. Change the mapping according to your needs, including any fields you want mapped outward to Microsoft 365 Defender.
7. Save your changes.

#### STEP 4 - Create an Incident in Microsoft 365 Defender  

For purposes of this use case, it can be a simple incident. The new incident will be ingested in Cortex XSOAR in approximately one minute.

#### STEP 5 - Add a Comment from Cortex XSOAR to Microsoft 365 Defender

In the example below, we have written *A comment from Cortex XSOAR to Microsoft 365 Defender*.

> 1. Create an entry in the incidents' war room.

Reviewer: ```suggestion
1. Create an entry in the incidents' War Room.
```

> 2. Click Actions > Tags and add the Comment Entry Tag To Microsoft 365 Defender tag.

Reviewer: ```suggestion
2. Click **Actions** > **Tags** and add the **Comment Entry Tag To Microsoft 365 Defender** tag.
```

| Microsoft365Defender.Incident.severity | string | Severity of the incident. Possible values are: UnSpecified, Informational, Low, Medium, and High. | 
| Microsoft365Defender.Incident.status | string | The current status of the incident. Possible values are: Active, Resolved, and Redirected. | 
| Microsoft365Defender.Incident.alerts | unknown | List of alerts relevant for the incidents. | 
| Microsoft365Defender.Incident.tags | unknown | List of custom tags associated with an incident, for example to flag a group of incidents with a common characteristic. | 
| Microsoft365Defender.Incident.comments | unknown | List of comments created by secops when managing the incident, for example additional information about the classification selection. | 

### microsoft-365-defender-incident-update

***
Update the incident with the given ID.

#### Base Command

`microsoft-365-defender-incident-update`

#### Input

| **Argument Name** | **Description** | **Required** |
| --- | --- | --- |

> | status | Categorize incidents (as Active, Resolved, or Redirected). Possible values are: Active, Resolved, Redirected, InProgress. | Optional | 

Reviewer: ```suggestion
| status | Categorize incidents. Possible values are: Active, Resolved, Redirected, InProgress. | Optional | 
```
Fix this is the yml file.

#### Base Command

`get-remote-data`

#### Input

| **Argument Name** | **Description** | **Required** |
| --- | --- | --- |
| id | The remote incident ID. | Required | 
| lastUpdate | The UTC timestamp in seconds of the last update. The incident is only updated if it was modified after the last update time. Default is 0. | Optional | 

#### Context Output

There is no context output for this command.

### get-mapping-fields

***
Returns the list of fields to map in outgoing mirroring. This command is only used for debugging purposes.

#### Base Command

`get-mapping-fields`

#### Input

| **Argument Name** | **Description** | **Required** |
| --- | --- | --- |

#### Context Output

There is no context output for this command.

### get-modified-remote-data

***
Get the list of incidents that were modified since the last update time. This method is used for debugging purposes. The get-modified-remote-data command is used as part of the Mirroring feature that was introduced in Cortex XSOAR version 6.1.

#### Base Command

`get-modified-remote-data`

#### Input

| **Argument Name** | **Description** | **Required** |
| --- | --- | --- |

> | lastUpdate | Date string representing the local time. The incident is only returned if it was modified after the last update time. | Optional | 

Reviewer: What is the format of the date string?

Author: any date string  "in a recognizably valid format"
